org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/BoundSet.java

```diff
@@ -615,12 +615,23 @@ boolean incorporate(InferenceContext18 context, TypeBound [] first, TypeBound []
 							}
 						}
 					}
+
 					if (deriveTypeArgumentConstraints) {
-						for (ConstraintTypeFormula typeArgumentConstraint : deriveTypeArgumentConstraints(bound1, bound2, context)) {
-							if (!reduceOneConstraint(context, typeArgumentConstraint))
-								return false;
+						// Try with a simpler (cheaper) approach first
+						for (ConstraintFormula cf : deriveTypeArgumentConstraintsSimple(bound1, bound2)) {
+							if (reduceOneConstraint(context, cf)) {
+								// No need to keep searching, the simple strategy worked
+								continue;
+							}
+
+							// If the simple search failed, try with a more complex (and expensive) search
+							for (ConstraintTypeFormula cf2 : deriveTypeArgumentConstraints(bound1, bound2, context)) {
```

Comment:
That looks like a for nested in a for, while the outer ConstraintFormula cf is not used. Feels wrong.

Reply:
Right, if the goal is to skip the inner loop when the simple approach was fully successful, then those should be two independent loops, where the second is run only if the first produced at least one FALSE.

Perhaps an empty result from `deriveTypeArgumentConstraintsSimple()` should also count as not successful and continue to the second loop.

Reply:
So for me this would mean that the `for`-loops need to remain nested, but before doing that I would like to know if you guys can make sense of this phenomenon.